crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -4524,6 +4524,13 @@ impl<'db> TypeInferenceBuilder<'db> {
                     Type::Unknown
                 }
             },
+            KnownInstanceType::Optional => {
+                let param_type = self.infer_annotation_expression(
+                    parameters,
+                    DeferredExpressionState::from(self.are_all_types_deferred()),
+                );
+                UnionType::from_elements(self.db, [param_type, Type::none(self.db)])
+            }
```

Comment:
I think `Optional` contains type expressions as per the grammar:
```
| <Optional> '[' type_expression ']'
```
So, you'll need to use `self.infer_type_expression` instead.

Reply:
Right but if I do that then I think Optional[Optional[int]] is not inferred correctly since we do not infer Optional out of annotations. I might be wrong but let me check. I think you are right anyway I will adjust this part.